## Imputation

### 1. vcf file converting

Genetype file should in [vcf](http://samtools.github.io/hts-specs/VCFv4.2.pdf) format, using ```./scripts/2vcf.py```to make this.

```python
python ./scripts/2vcf.py -r ./data/ASA_CHIA_ANNO_FINAL_1125.txt -i ./data/genotype.txt -o ./data/genotype.vcf -n 10
```

*notes:* 

* ALT could be same as REF
* locas could duplicate

### 2. quality control

Firstly, convert vcf file to binary plink format

```bash
plink2 --vcf genotype.vcf --allow-extra-chr --vcf-half-call 'm' --make-bed --out genotype
```

The './0' is specified as an error in plink, using --vcf-half-call 'm' to treat halfs as missing.

some R [scripts](https://github.com/MareesAT/GWA_tutorial)

#### 2.1 missingness 

plot histogram of missing rate with:

```bash
plink --bfile ../genotype --missing
Rscript --no-save ../../../scripts/hist_miss.R
```

```bash
# Delete SNPs with missingness >0.2
plink --bfile ../genotype --geno 0.2 --make-bed --out genotype_1
# Delete individuals with missingness >0.2
plink --bfile genotype_1 --mind 0.2 --make-bed --out genotype_2
# Delete SNPs with missingness >0.05
plink --bfile genotype_2 --geno 0.05 --make-bed --out genotype_3
```

![](./data/QC/missing/histlmiss.png)

![](./data/QC/missing/histimiss.png)

#### 2.2 sex discrepancy

Subjects who were a priori determined as females must have a **F value of <0.2**, and subjects who were a priori determined as males must have a **F value >0.8**. This F value is based on the X chromosome inbreeding (homozygosity) estimate. Subjects who do not fulfil these requirements are flagged "**PROBLEM**" by PLINK.

```bash
plink --bfile ../missing/genotype_3 --check-sex 
```

...

#### 2.3 minor allele frequency (MAF)

Generate a bfile with autosomal SNPs only.

```bash
awk '{ if ($1 >= 1 && $1 <= 22) print $2 }' ../missing/genotype_3.bim > snp_1_22.txt
plink --bfile ../missing/genotype_3 --extract snp_1_22.txt --make-bed --out genotype_4
```

check MAF distribution

```bash
plink --bfile genotype_4 --freq --out MAF_check
Rscript --no-save ../../../scripts/MAF_check.R
```

flitering

```bash
plink --bfile genotype_4 --maf 0.06 --make-bed --out genotype_5
```

![](./data/QC/maf/MAF_distribution.png)

#### 2.4 Hardy-Weinberg equilibrium (HWE)

Check the distribution of HWE p-values of all SNPs.

```bash
plink --bfile ../maf/genotype_5 --hardy
```

Selecting SNPs with HWE p-value below 0.00001, required for one of the two plot generated by the next Rscript, allows to zoom in on strongly deviating SNPs. 

```bash
awk '{ if ($9 <0.00001) print $0 }' plink.hwe>plinkzoomhwe.hwe
```

By default the --hwe option in plink only filters for controls

```bash
plink --bfile ../maf/genotype_5 --hwe 0.01 --make-bed --out genotype_filter_step1
```

This second HWE step only focusses on cases because in the controls all SNPs with a HWE p-value < hwe 1e-6 were already removed

```bash
 plink --bfile genotype_filter_step1 --hwe 0.05 --hwe-all --make-bed --out genotype_6
```

![](./data/QC/hwe/histhwe.png)

#### 2.5 heterozygosity

Cheack the heterozygosity rate

```bash
plink --bfile ../hwe/genotype_6 --indep-pairwise 50 5 0.2 --out indepSNP
plink --bfile ../hwe/genotype_6 --extract indepSNP.prune.in --het --out R_check
Rscript --no-save ../../../scripts/check_heterozygosity_rate.R 
```

![](./data/QC/het/heterozygosity.png)

With the offered R script, the individuals with a heterozygosity rate deviating more than 3 sd from the mean will be selected, the generated file is than adapted to plink.

```bash
Rscript --no-save ../../../scripts/heterozygosity_outliers_list.R
sed 's/"// g' fail-het-qc.txt | awk '{print$1, $2}'> het_fail_ind.txt
```

Remove heterozygosity rate outliers.

```bash
plink --bfile ../hwe/genotype_6 --remove het_fail_ind.txt --make-bed --out genotype_7
```

#### 2.6 cryptic relatedness

......

### 3. phasing

using beagle

```bash
java -jar ../scripts/beagle.18May20.d20.jar gt=../data/genotype.vcf out=../data/genotype.gt
```

### 4. preparing

Split the phased genotype file by chromosome with ```./scripts/merge_and_split.py```

```bash
python3 ./scripts/merge_and_split.py -op s -i ./data/genotype.gt.vcf.gz -o ./data/vcfs
```

Download the [reference](http://bochet.gcc.biostat.washington.edu/beagle/1000_Genomes_phase3_v5a/) and adjust the target genomic position consistent with the reference with [conform-gt](http://faculty.washington.edu/browning/conform-gt.html). The [population](http://bochet.gcc.biostat.washington.edu/beagle/1000_Genomes_phase3_v5a/sample_info/integrated_call_samples_v3.20130502.ALL.panel) infomation of reference genotype is listed below:

|  super_pop  | abbreviation |
| :---------: | :----------: |
| East Asian  |     EAS      |
| South Asian |     SAS      |
|   African   |     AFR      |
|  European   |     EUR      |
|  American   |     AMR      |

conform-gt usage:

```bash
ls /data/share/1KG/b37/b37.bref3 | cut -d '.' -f1 | while read line; do java -jar scripts/conform-gt.24May16.cee.jar ref=/data/share/1KG/b37/b37.vcf/chr${line:3}.1kg.phase3.v5a.b37.vcf.gz gt=./data/vcfs/chr${line:3}.vcf.gz chrom=${line:3} out=./data/vcfs/conform.chr${line:3}; done
```

Conversion between vcf and bref3 can be achived with [bref3](http://faculty.washington.edu/browning/beagle/bref3.18May20.d20.jar) and [unbref3](http://faculty.washington.edu/browning/beagle/unbref3.18May20.d20.jar) as below:

```bash
ls * | cut -d '.' -f 1-5 | while read line; do java -jar ~/Imputation/scripts/unbref3.18May20.d20.jar $line.bref3 > ../b37.vcf/$line.vcf; done
```

### 5. Imputation 

Impute with Beagle5.1

```bash
ls ./data/vcfs/conform.* | cut -d '.' -f3 | uniq | while read line; do java -jar ./scripts/beagle.18May20.d20.jar ref=/data/share/1KG/b37/b37.bref3/$line.1kg.phase3.v5a.b37.bref3 gt=./data/vcfs/conform.$line.vcf.gz out=./data/imputed/$line; done
```

Merge file

```bash
python3 ./scripts/merge_and_split.py -op m -i ./data/imputed -o ./data/genotype.imputed
```











